tests/DCPS/AddressLookup/main.cpp

```diff
@@ -0,0 +1,160 @@
+  unsigned short port_number = 0;
+
+  addrinfo hints;
+  std::memset(&hints, 0, sizeof(hints));
+#ifdef ACE_HAS_IPV6
+  hints.ai_family = AF_UNSPEC;
+#else
+  hints.ai_family = AF_INET;
+#endif
+
+  // The ai_flags used to contain AI_ADDRCONFIG as well but that prevented
+  // lookups from completing if there is no, or only a loopback, IPv6
+  // interface configured. See Bugzilla 4211 for more info.
+
+#if defined ACE_HAS_IPV6 && !defined IPV6_V6ONLY
+  hints.ai_flags |= AI_V4MAPPED;
+#endif
+
+#if defined ACE_HAS_IPV6 && defined AI_ALL
+  // Without AI_ALL, Windows machines exhibit inconsistent behaviors on
+  // difference machines we have tested.
+  hints.ai_flags |= AI_ALL;
+#endif
+
+  // Note - specify the socktype here to avoid getting multiple entries
+  // returned with the same address for different socket types or
+  // protocols. If this causes a problem for some reason (an address that's
+  // available for TCP but not UDP, or vice-versa) this will need to change
+  // back to unrestricted hints and weed out the duplicate addresses by
+  // searching this->inet_addrs_ which would slow things down.
+  hints.ai_socktype = SOCK_STREAM;
+
+  addrinfo *res = 0;
+  const int error = ACE_OS::getaddrinfo(address.c_str(), 0, &hints, &res);
+  if (error) {
+    ACE_ERROR((LM_ERROR, "hostname_to_ip: Call to getaddrinfo() for hostname %C returned error: %d\n", address.c_str(), gai_strerror(error)));
+    return;
+  }
+
+  for (addrinfo* curr = res; curr; curr = curr->ai_next) {
+    if (curr->ai_family != AF_INET && curr->ai_family != AF_INET6) {
+      ACE_DEBUG((LM_DEBUG, "hostname_to_ip: Encounter an address that is not AF_INET or AF_INET6\n"));
+      continue;
+    }
+
+    union ip46 {
+        sockaddr_in  in4_;
+#ifdef ACE_HAS_IPV6
+        sockaddr_in6 in6_;
+#endif /* ACE_HAS_IPV6 */
+    } addr;
+    std::memset(&addr, 0, sizeof(addr));
+
+#ifdef ACE_HAS_IPV6
+    ACE_DEBUG((LM_DEBUG, "ip46.in6_ size is %d\n", sizeof(addr.in6_)));
+#endif
+    ACE_DEBUG((LM_DEBUG, "addr size is %d, curr addrinfo size is %d, curr->ai_addrlen is %d, curr->ai_addr size is %d, curr family is %d\n",
+               sizeof(addr), sizeof(*curr), curr->ai_addrlen, sizeof(*(curr->ai_addr)), curr->ai_family));
+    std::memcpy(&addr, curr->ai_addr, curr->ai_addrlen);
+    //std::memcpy(&addr, curr->ai_addr, sizeof addr);
+
+#ifdef ACE_HAS_IPV6
+    if (curr->ai_family == AF_INET6) {
+      addr.in6_.sin6_port = ACE_NTOHS(port_number);
+    } else {
+#endif /* ACE_HAS_IPV6 */
+      addr.in4_.sin_port = ACE_NTOHS(port_number);
+#ifdef ACE_HAS_IPV6
+    }
+#endif /* ACE_HAS_IPV6 */
+
+    ACE_INET_Addr temp;
+    temp.set_addr(&addr, sizeof(addr));
+    temp.set_port_number(port_number, 1 /*encode*/);
+
+    print_addr(temp, "==== IP address:");
+  }
+  ACE_OS::freeaddrinfo(res);
+}
+
+void address_info() {
+  size_t addr_count;
+  ACE_INET_Addr *addr_array = 0;
```

Comment:
An ASAN build complains about leaking this array. So we probably want to add something like these from NetworkResource.cpp: https://github.com/objectcomputing/OpenDDS/blob/master/dds/DCPS/NetworkResource.cpp#L89-L101